File: src/year2024/day3/python/part1.py

# lines = open("../test1.txt").readlines()
# lines = open("../prod1.txt").readlines()
# lines = open("../test2.txt").readlines()
lines = open("../prod2.txt").readlines()

p = "mul("

ans = 0
for line in lines:
    for i in range(len(line)):
        j = i

        if line[i:].startswith("mul("):
            while line[j] != ")":
                j += 1
            new_line = line[i:j+1].replace("mul(", "").replace(")","")

            try:
                l,r = new_line.split(",")
                ans += int(l) * int(r)
            except:
                pass

print(ans)

File: src/year2024/day3/scala/part1.scala

@main def part1(): Int = {
  val lines = scala.io.Source.fromFile("../prod1.txt").getLines().toList
  var ans: Int = 0

  for (line <- lines) {
    val l_length: Int = line.length
    for (i <- 0 until line.length) {
      if (line.slice(i, l_length).startsWith("mul(")) {
        var end: Int = i
        while (line.charAt(end) != ')') {
          end += 1
        }

        val new_slice = line.slice(i, end+1).replace("mul(", "").replace(")", "")
        var splits = new_slice.split(",")
        // NOTE: could also do it with scala.util.Try -> `Try(splits(0).toInt).isSuccess == Boolean`
        if (splits.length == 2 && splits(0).toIntOption.isDefined && splits(1).toIntOption.isDefined) {
              val n1 = splits(0).toInt
              val n2 = splits(1).toInt
              ans += (n1 * n2)
        }

      }
    }
  }

  ans
}
